### ISSUE TITLE
[CI] Cluster Mesh upgrade: `no-interrupted-connections`: `test-conn-disrupt-client` failed due to interrupted traffic during downgrade

### ISSUE THREAD

**Issue by qmonnet** [cilium/cilium#30964]
Over the last month, we have been observing several occurrences of the **Cilium Cluster Mesh upgrade** workflow failing the connectivity check `no-interrupted-connections`, on Pull Requests against the `main` branch. We have at least 4 reports so far:

- https://github.com/cilium/cilium/actions/runs/7712912167/job/21021566215 (#30528)
    - 2024-01-30
    - job config: `1, disabled, iptables, false`
    - step: post-downgrade to 1.15.0-rc.1
    - `Pod test-conn-disrupt-client-599598f4f9-d4bhq flow was interrupted (restart count does not match 0 != 1)`

- https://github.com/cilium/cilium/actions/runs/7898508466/job/21556052907 (#30732)
    - 2024-02-14
    - job config: `2, disabled, none, false`
    - step: post-downgrade to 1.15.0
    - `Pod test-conn-disrupt-client-599598f4f9-xm5xs flow was interrupted (restart count does not match 0 != 1)`

- https://github.com/cilium/cilium/actions/runs/8016451099/job/21898409307 (#30885)
    - 2024-02-23
    - job config: `1, disabled, iptables, false, 255`
    - step: post-downgrade to 1.15.1
    - `Pod test-conn-disrupt-client-599598f4f9-6dp8g flow was interrupted (restart count does not match 0 != 1)`
    - [cilium-sysdumps.zip.7z.zip](https://github.com/cilium/cilium/files/14403464/cilium-sysdumps.zip.7z.zip) (those from the previous reports have expired)

- https://github.com/cilium/cilium/actions/runs/8049142040/job/21981914884 (#30742)
    - 2024-02-26
    - job config: `4, wireguard, iptables, false, 511`
    - step: post-downgrade to 1.15.1
    - `Pod test-conn-disrupt-client-599598f4f9-fr9rm flow was interrupted (restart count does not match 0 != 1)`
    - [cilium-sysdumps.zip.7z.zip](https://github.com/cilium/cilium/files/14406113/cilium-sysdumps.zip.7z.zip)

_Originally posted by @qmonnet in https://github.com/cilium/cilium/issues/28088#issuecomment-1963811672_. This other issue is related in the sense that it tracked `no-interrupted-connections` failures as well. The affected workflow were different and it's likely that the root cause is distinct here, however, readers might refer to the former issue for hints on how to debug (for example using pwru).

If you observe similar symptoms on a workflow different from Cluster Mesh upgrade, please open a separate issue.

**Comment by bimmlerd** [comment]
The `statedb/reflector: fix race condition in test` hit is another on `4, wireguard, iptables, false, 511`: https://github.com/cilium/cilium/actions/runs/8051125416/job/21988310944#step:42:53

**Comment by pippolo84** [comment]
Another one [here](https://github.com/cilium/cilium/actions/runs/8083701904/job/22087347247) from https://github.com/cilium/cilium/pull/30766

Job config: `(4, wireguard, iptables, false, 511)`
Full Logs: [logs_21213476335.zip](https://github.com/cilium/cilium/files/14439430/logs_21213476335.zip)
Sysdump: [cilium-sysdumps (19).zip](https://github.com/cilium/cilium/files/14439419/cilium-sysdumps.19.zip)

**Comment by bimmlerd** [comment]
Hit on `(2, disabled, none, false, 511)` in https://github.com/cilium/cilium/actions/runs/8140815007/job/22246772143#step:42:54
sysdump: 
[cilium-sysdump-ipsec-downgrade-1-20240305-085616.7z.zip](https://github.com/cilium/cilium/files/14496240/cilium-sysdump-ipsec-downgrade-1-20240305-085616.7z.zip) (sorry for the stupid format, hard to get these under 25M)

**Comment by aditighag** [comment]
Hit on - https://github.com/cilium/cilium/actions/runs/8236859840/job/22524289481.
Sysdump 


[cilium-sysdumps (4).zip](https://github.com/cilium/cilium/files/14562485/cilium-sysdumps.4.zip)

**Comment by giorio94** [comment]
I did some more testing, and this turned out to be a real issue.

Essentially, it manifests during the downgrade step, which additionally disables kvstoremesh at the same time. More specifically, it appears to affect the agents which have not yet been restarted when kvstoremesh is being terminated, whose etcd watchers fail with the following error:

```
2024-04-12T12:23:46.166818116Z {"level":"warn","ts":"2024-04-12T12:23:46.165965Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0015ee380/clustermesh-apiserver.kube-system.svc:2379","attempt":0,"error":"rpc error: code = Unknown desc = malformed header: missing HTTP content-type"}
```

It is not yet totally clear to me the cause of the error, although it is likely related with the reconnection of the watchers to a different clustermesh-apiserver instance. The error triggers [this logic](https://github.com/cilium/cilium/blob/726a93b559f91d0dac11bd234437fb2005aadcb6/pkg/kvstore/etcd.go#L920-L934) in the Cilium etcd client, in turn triggering a full relist of the prefix. However, being a totally different etcd instance which only contains local information, that causes the agent to delete all the entries corresponding to the remote cluster.

This issue has already been addressed on main (https://github.com/cilium/cilium/pull/31677), which ensures that the watchers don't get resumed if reconnecting to a different etcd instance, but instead triggers a full reconnection to guarantee consistency.

I'll do a bit more testing to confirm the exact cause of the error  (and why it doesn't manifest in other cases), but my feeling is that we may need to implement a reduced version of that patch for stable versions as well to fully address this flake.

/cc @thorn3r, as the author of 31677
/cc @bimmlerd, as we brainstormed offline about this flake

**Comment by giorio94** [comment]
That had been my second guess after ruling out the resource starvation theory, but I struggled a bit to prove it, as the deletion happens on the agents before being restarted, which means that neither logs nor maps contents are captured in the sysdumps.

I also kinda knew that the reconnection to a different etcd instance could be potentially problematic if the watch restarted, but I never really observed that to happen in practice (and especially not due to that error, which causes the full relist).

That said, I modified the workflow to immediately capture the sysdump after the crash of the conn-disrupt pod, and the logs effectively mentioned the generation of a removal event for that IP. Similarly, the ipcache map didn't contain the corresponding entries anymore.

**Comment by thorn3r** [comment]
Excellent detective skills :detective:. How limited are you thinking? Backporting https://github.com/cilium/cilium/pull/31677/commits/97033e9d2385c028adb5e71c15fcf5e061f33f06 and https://github.com/cilium/cilium/pull/31677/commits/b2e9eb65892038878152118044c66843ba602d4c should address that. But at that point we might as well include https://github.com/cilium/cilium/pull/31677/commits/bbf448878038ddff8d484dd1704820df7962bc4e as well to have coverage, since it's not specific to HA and tests the handling of reconnections

**Comment by giorio94** [comment]
Opened https://github.com/cilium/cilium/pull/32005 to backport a reduced version of #31677 to v1.15.

**Comment by giorio94** [comment]
The fix has been merged into v1.15 as well. Looking at the main branch, I haven't seen any related failures since https://github.com/cilium/cilium/pull/31677 got merged (~64 runs). I'm going to close this issue then, feel free to reopen if you see any new occurrences.
